### ISSUE TITLE
Wishlist/Suggestion: Return the Request object on send error

### ISSUE THREAD

**Issue by ragnese** [algesten/ureq#333]
I realize this might be a pretty invasive change, but I'd find it very helpful if ureq::Error contained the Request that failed to send.

I think this is reasonable because the Request is actually mutated as part of several of the `send()` methods. So we can't just clone the Request before we send it, because that wont be the true Request that was transmitted. Also, that would be frustrating because you'd clone it for the error case, but if everything succeeds, you made a clone for no reason.

So, basically, I'm requesting that ureq::Error goes from:
```
pub enum Error {
    Status(u16, Response),
    Transport(Transport),
}
```

to something like:
```
pub enum ErrorReason {
    Status(u16, Response),
    Transport(Transport),
}

pub struct Error {
    pub request: Request,
    pub reason: ErrorReason,
}
```

Or something similar.

Thank you for sharing ureq with us. I really appreciate it.

Cheers,
Rob

**Comment by jsha** [comment]
An interesting idea! One thing to note is that `Request` is actually a builder-style object. At `send()`, we build it into an internal-only object (currently called `Unit`, but could just as well be called `BuiltRequest` or `ReadyRequest`). That built object is probably what you want in terms of what _really_ got sent on the wire. Also, it might be more accurate to share a series of `BuiltRequest` / `Response` pairs (for redirects). We would presumably want to box these up if we do it, to avoid ballooning the size of `Error`. Also, we likely wouldn't do it exactly the way you propose, since that would break our Error API. Another way to do it would be to add a `BuiltRequest` field on `Response`, and also on `Transport`. Then we would expose a method for `Error` that would return the appropriate thing.

Can you tell us a bit more about your use case? What would having a copy of the sent request allow you to do that you can't do today?

**Comment by ragnese** [comment]
Good points. I did peek at the code and realized that you don't technically add those fields to Request- you deconstruct the actual Request and build this Unit thing, as you described. But the data is still all there- you probably *could* expose it as a Request rather than a new ReadyRequest or whatever, if that makes for a nicer API.

I haven't given any thought to redirects. I guess it does make sense to give the whole chain of Request-Response pairs for the sake of being able to see what *really* happened in the process.

Certainly having the ability to get the request from Response or Transport would be fine. I guess you can't even add it to each variant of the Error- that would still break the API. Oh well.

My specific use case is that I'm writing some code for a fleet of devices that are not guaranteed to have stable/good internet connections. So, failed requests are totally fine and expected. But rather than holding it in memory and retrying, we serialize the request to a local cache database and the requests can be "replayed" at a later time (even after loss of power, etc), by another process.

So, rather than assuming that the request will always be sent with ureq (which is true today, but might not always be true and certainly wasn't true before I switched to ureq last week ;) ), I have chosen to serialize the request as literally as possible, which means including all headers.

Plus, it just seems nice from a logging POV as well. It would be nice to log the "true" request as the server would have seen it- especially if it was rejected for some reason.

**Comment by algesten** [comment]
Closing since we're moving to ureq 3.x. The problem remains though.
